src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/IntervalTreeHelpers.cs

```diff
@@ -0,0 +1,207 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Shared.Collections;
+
+internal interface IIntervalTreeHelper<T, TIntervalTree, TNode>
+    where TIntervalTree : IIntervalTree<T>
+{
+    public bool TryGetRoot(TIntervalTree tree, [NotNullWhen(true)] out TNode? root);
+    public bool TryGetLeftNode(TIntervalTree tree, TNode node, [NotNullWhen(true)] out TNode? leftNode);
+    public bool TryGetRightNode(TIntervalTree tree, TNode node, [NotNullWhen(true)] out TNode? rightNode);
+
+    public T GetValue(TIntervalTree tree, TNode node);
+    public TNode GetMaxEndNode(TIntervalTree tree, TNode node);
+}
+
+/// <summary>
+/// Utility helpers used to allow code sharing for the different implementations of <see cref="IIntervalTree{T}"/>s.
+/// </summary>
+internal static class IntervalTreeHelpers<T, TIntervalTree, TNode, TIntervalTreeHelper>
+    where TIntervalTree : IIntervalTree<T>
+    where TIntervalTreeHelper : struct, IIntervalTreeHelper<T, TIntervalTree, TNode>
+{
+    private static readonly ObjectPool<Stack<TNode>> s_nodeStackPool = new(() => new(), 128, trimOnFree: false);
+
+    public static IEnumerator<T> GetEnumerator(TIntervalTree tree)
+    {
+        var helper = default(TIntervalTreeHelper);
+        if (!helper.TryGetRoot(tree, out var root))
+            return SpecializedCollections.EmptyEnumerator<T>();
+
+        return GetEnumeratorWorker(tree, helper, root);
+
+        static IEnumerator<T> GetEnumeratorWorker(
+            TIntervalTree tree, TIntervalTreeHelper helper, TNode root)
+        {
+            using var _ = s_nodeStackPool.GetPooledObject(out var stack);
+            var current = (Node: root, HasValue: true);
```

Comment:
>var current = (Node: root, HasValue: true); [](http://example.com/codeflow?start=12&length=43)

can you expand this out to two separate variables, rather than the valuetuple?

Reply:
sure.

Reply:
done!